fabric-structure-events-v0/src/main/java/net/fabricmc/fabric/mixin/event/structure/StructureStartMixin.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.event.structure;
+
+import java.util.Random;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import net.minecraft.structure.StructurePiece;
+import net.minecraft.structure.StructureStart;
+import net.minecraft.util.math.BlockBox;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.StructureWorldAccess;
+import net.minecraft.world.gen.StructureAccessor;
+import net.minecraft.world.gen.chunk.ChunkGenerator;
+
+import net.fabricmc.fabric.api.event.structure.v0.StructureFeatureEvents;
+import net.fabricmc.fabric.api.event.structure.v0.StructurePieceEvents;
+
+@Mixin(StructureStart.class)
+public class StructureStartMixin {
+	@Redirect(method = "generateStructure", at = @At(value = "INVOKE", target = "Lnet/minecraft/structure/StructurePiece;generate(Lnet/minecraft/world/StructureWorldAccess;Lnet/minecraft/world/gen/StructureAccessor;Lnet/minecraft/world/gen/chunk/ChunkGenerator;Ljava/util/Random;Lnet/minecraft/util/math/BlockBox;Lnet/minecraft/util/math/ChunkPos;Lnet/minecraft/util/math/BlockPos;)Z"))
+	private boolean onStructurePieceAdded(StructurePiece structurePiece, StructureWorldAccess structureWorldAccess, StructureAccessor structureAccessor, ChunkGenerator chunkGenerator, Random random, BlockBox boundingBox, ChunkPos chunkPos, BlockPos blockPos) {
+		StructurePieceEvents.PIECE_ADDED.invoker().onStructurePieceAdded(structurePiece, structureWorldAccess.toServerWorld());
```

Comment:
So this is called just before the piece is generated. My concern here is someone uses this event to inject their own stuff into existing structures. The big issue with this is you call `toServerWorld` which provides the server world the chunks will eventually be generated to.

However directly calling `setBlockState` on the server world in this state will deadlock the game.

Reply:
I now expose a `StructureWorldAccess` instead of a `ServerWorld`. Granted, users can still call `toServerWorld` themselves, but this along with the JavaDoc explaining it might be enough to prevent people from creating deadlocks.